### PR TITLE
consolidate string generation code

### PR DESCRIPTION
**Pull request by WalterBright**
The idea is to consolidate the string emission code into one place, where the string caching code can then be fixed.

### CODE REVIEW

**Comment by yebblies** [comment]
Do you really need `toStringDarraySymbol`?  Wouldn't it be better to generate a pair instead of writing it out to the data segement?

**Comment by WalterBright** [comment]
Sometimes you need one, sometimes the other.

**Comment by yebblies** [comment]
Fair enough.

Do you know how to fix https://issues.dlang.org/show_bug.cgi?id=13957 ?  It's a pretty serious bug and outside my level of backend knowledge.

**Comment by WalterBright** [comment]
Well, I'll have to learn how to fix it. You're right that it's awfully tricky.

**Comment by yebblies** [comment]
I got pretty close, it managed to allocate registers correctly, but I don't think I was using the right elems for passing the struct as a parameter.  I'm hoping it's just a little bit of missing support, that was previously masked by 16-byte structs always getting converted to ucent/cdouble.

**Comment by yebblies** [comment]
Auto-merge toggled on

**Comment by WalterBright** [comment]
thanks for the quick action!

**Comment by yebblies** [comment]
What made the previous commit segfault?

**Comment by WalterBright** [comment]
put out the wrong string length

File: src/e2ir.c [hunk 1]
```diff
@@ -5353,3 +5326,41 @@ elem *toElemDtor(Expression *e, IRState *irs)
     er = appendDtors(irs, er, starti, endi);
     return er;
 }
+
+
+/*******************************************************
+ * Write read-only string to object file, create a local symbol for it.
+ * str[len] must be 0.
```
Comment (braddr) [inline src/e2ir.c]: Why isn't this requirement expressed in the form of an assertion?

File: src/e2ir.c [hunk 2]
```diff
@@ -55,6 +55,8 @@ unsigned totym(Type *tx);
 Symbol *toSymbol(Dsymbol *s);
 elem *toElem(Expression *e, IRState *irs);
 dt_t **Expression_toDt(Expression *e, dt_t **pdt);
+Symbol *toStringSymbol(const char *str, size_t len, size_t sz);
+Symbol *toStringDarraySymbol(const char *str, size_t len, size_t sz);
```
Comment (braddr) [inline src/e2ir.c]: Shouldn't these be declared in a header rather than manual re-declaration?

Comment (yebblies) [inline src/e2ir.c]: There isn't a header set up for the glue layer stuff, and it doesn't belong in any of the frontend headers.

**Pull request merged**